aten/src/ATen/core/function_schema.h

```diff
@@ -327,6 +339,72 @@ struct FunctionSchema {
     }
   }
 
+  void flattenAliasTypeSet(const ArrayRef<TypePtr>& types, std::unordered_set<TypePtr>& flattened_types) const {
+    for (const TypePtr& type: types) {
+      if (!flattened_types.count(type)) {
+        flattenAliasTypeSet(type->containedTypes(), flattened_types);
+        flattened_types.insert(type);
+      }
+    }
+  }
+
+  c10::optional<std::vector<TypePtr>> mapTypeToAliasTypeSet(const TypePtr& type) const {
+    switch(type->kind()) {
+      case TypeKind::ListType:
+      case TypeKind::DictType:
+      case TypeKind::ClassType:
+      case TypeKind::TensorType:
+        return std::vector<TypePtr> {c10::unshapedType(type)};
+      case TypeKind::UnionType: {
+        std::vector<TypePtr> mutable_types;
+        for (const TypePtr& inner :
+             type->expectRef<UnionType>().containedTypes()) {
+          if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+            mutable_types.insert(
+                mutable_types.end(),
+                (*maybe_inner_types).begin(),
+                (*maybe_inner_types).end());
+          }
+        }
+        if (mutable_types.size() == 0) {
+          return c10::nullopt;
+        }
+        return mutable_types;
+      }
+      case TypeKind::AnyType:
+        return {std::vector<TypePtr>{type}};
+      case TypeKind::OptionalType: {
+        auto inner = type->castRaw<OptionalType>()->getElementType();
+        return mapTypeToAliasTypeSet(inner);
+      }
+      case TypeKind::TupleType: {
+        std::vector<TypePtr> mutable_types;
+        for (const TypePtr& inner : type->expectRef<TupleType>().elements()) {
+          if (auto maybe_inner_types = mapTypeToAliasTypeSet(inner)) {
+            mutable_types.insert(
+                mutable_types.end(),
+                (*maybe_inner_types).begin(),
+                (*maybe_inner_types).end());
+          }
+        }
+        if (mutable_types.size() == 0) {
+          return c10::nullopt;
+        }
+        return {std::vector<TypePtr>{TupleType::create(mutable_types)}};
+      }
+      default:
+        return c10::nullopt;
+    }
+  }
```

Comment:
Put the implementation functions like this in `function_schema.cpp` instead of `function_schema.h`